src/services/section-helper.js

```diff
@@ -0,0 +1,112 @@
+import deepEqual from 'fast-deep-equal';
+import Logger from '../common/logger.js';
+import uniqueId from '../utils/unique-id.js';
+import cloneDeep from '../utils/clone-deep.js';
+
+const logger = new Logger(import.meta.url);
+
+function checkPlausibility({ section, ancestorSection = null }) {
+  if (!section.key || !section.type) {
+    throw new Error('Section key and type must be specified');
+  }
+
+  if (ancestorSection && (!ancestorSection.key || !ancestorSection.type)) {
+    throw new Error('Ancestor section key and type must be specified');
+  }
+
+  if (ancestorSection && ancestorSection.type !== section.type) {
+    throw new Error(`Ancestor section has type ${ancestorSection.type} and cannot be changed to ${section.type}`);
+  }
+}
+
+export function createSectionRevision({ section, ancestorSection = null, isRestoreOperation = false }) {
+  checkPlausibility({ section, ancestorSection });
+
+  // If the section has no content, it could be ...
+  if (!section.content) {
+
+    // (a) ... the restore of a hard-deleted section
+    if (isRestoreOperation) {
+
+      // In this case the date, user and reason have to be defined:
+      if (!section.deletedOn || !section.deletedBy || !section.deletedBecause) {
+        throw new Error('Restoring a deleted section without deletion information is not allowed');
+      }
+
+      // If the deletion information is exactly the same as in the ancestor, we continue the revision:
+      if (
+        section.deletedOn.getTime() === ancestorSection?.deletedOn?.getTime()
+        && section.deletedBy === ancestorSection?.deletedBy
+        && section.deletedBecause === ancestorSection?.deletedBecause
+      ) {
+        logger.info(`Section has not changed compared to ancestor section with revision ${ancestorSection.revision}, using the existing`);
+        return cloneDeep(ancestorSection);
+      }
+
+      // Otherwise we start a new revision:
+      logger.info(`Creating new revision for section key ${section.key}`);
+      return {
+        revision: uniqueId.create(),
+        key: section.key,
+        deletedOn: cloneDeep(section.deletedOn),
+        deletedBy: cloneDeep(section.deletedBy),
```

Comment:
These values don't need cloning, right?

Reply:
Not the `deletedBy`, and for `deletedOn` we can only take `new Date(section.deletedOn.getTime())`, will do.

Reply:
Done